Grails 6.2.2: ClassCastException for inheritance in Commands

### Expected Behavior

Grails throws exception ClassCastException when we have 2 endpoints with inherited commands. It worked fine on 6.1.2.

Looks like this is a critical issue.

```
class UrlMappings {
        "/my/$patientId/procedure"(controller: "my", action: "createLedgerProcedure", method: "POST")
        "/my/$patientId/procedure/insuranceEstimates"(controller: "my", action: "getInsuranceEstimates", method: "POST")
}

class MyBaseCommand implements Validateable {

    int testId
}

class MyCommand extends MyBaseCommand {

    int myId
}

class MyController {
    def getInsuranceEstimates(MyCommand command, long patientId) {
        log.info('test2 method')
        render([response: "test2 method"] as JSON)
    }

    def createLedgerProcedure(MyBaseCommand command, long patientId) {
        log.info('test1 method')
        render([response: "test1 method"] as JSON)
    }

}
```



### Actual Behaviour

Exception:
```
2024-12-26 18:13:10.639 ERROR --- [nio-8080-exec-2] StackTrace                               : Full Stack Trace:

java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:211)
	at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
	at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1072)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:965)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassCastException: class com.example.MyBaseCommand cannot be cast to class com.example.MyCommand (com.example.MyBaseCommand and com.example.MyCommand are in unnamed module of loader org.springframework.boot.devtools.restart.classloader.RestartClassLoader @8f4db8)
	... 13 common frames omitted

2024-12-26 18:13:10.690 ERROR --- [nio-8080-exec-2] o.g.web.errors.GrailsExceptionResolver   : ClassCastException occurred when processing request: [POST] /my/123/procedure
class com.example.MyBaseCommand cannot be cast to class com.example.MyCommand (com.example.MyBaseCommand and com.example.MyCommand are in unnamed module of loader org.springframework.boot.devtools.restart.classloader.RestartClassLoader @8f4db8). Stacktrace follows:

java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:211)
	at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
	at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1072)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:965)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassCastException: class com.example.MyBaseCommand cannot be cast to class com.example.MyCommand (com.example.MyBaseCommand and com.example.MyCommand are in unnamed module of loader org.springframework.boot.devtools.restart.classloader.RestartClassLoader @8f4db8)
	... 13 common frames omitted

```

### Steps To Reproduce

Reproduces on Grails 6.2.2. It worked fine on 6.1.2.

[grails_6_2_2.zip](https://github.com/user-attachments/files/18254075/grails_6_2_2.zip)


An example of the application has been attached:
- run the attached application: `./gradlew bootRun`
- Send the request:
```

curl --location --request POST 'http://localhost:8080/my/123/procedure' \
--header 'Content-Type: application/json' \
--data '{
    "testId": 321
}'
```

### Environment Information

- Windows 10
- Java:
```
openjdk version "11.0.22" 2024-01-16
OpenJDK Runtime Environment Temurin-11.0.22+7 (build 11.0.22+7)
OpenJDK 64-Bit Server VM Temurin-11.0.22+7 (build 11.0.22+7, mixed mode)
```

### Example Application

_No response_

### Version

6.2.2

--- a/grails-plugin-controllers/src/main/groovy/org/grails/compiler/web/ControllerActionTransformer.java
+++ b/grails-plugin-controllers/src/main/groovy/org/grails/compiler/web/ControllerActionTransformer.java
@@ -90,11 +90,23 @@ import java.io.File;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
 import java.net.URL;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 import java.util.regex.Pattern;
 
 import static org.codehaus.groovy.ast.tools.GeneralUtils.*;
-import static org.grails.compiler.injection.GrailsASTUtils.*;
+import static org.grails.compiler.injection.GrailsASTUtils.applyDefaultMethodTarget;
+import static org.grails.compiler.injection.GrailsASTUtils.applyMethodTarget;
+import static org.grails.compiler.injection.GrailsASTUtils.buildGetMapExpression;
+import static org.grails.compiler.injection.GrailsASTUtils.buildGetPropertyExpression;
+import static org.grails.compiler.injection.GrailsASTUtils.buildSetPropertyExpression;
+import static org.grails.compiler.injection.GrailsASTUtils.hasAnnotation;
+import static org.grails.compiler.injection.GrailsASTUtils.hasParameters;
+import static org.grails.compiler.injection.GrailsASTUtils.isInheritedFromTrait;
+import static org.grails.compiler.injection.GrailsASTUtils.removeAnnotation;
 
 /**
  * Enhances controller classes by converting closures actions to method actions and binding

--- a/grails-plugin-controllers/src/main/groovy/org/grails/compiler/web/ControllerActionTransformer.java
+++ b/grails-plugin-controllers/src/main/groovy/org/grails/compiler/web/ControllerActionTransformer.java
@@ -90,23 +90,11 @@ import java.io.File;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
 import java.net.URL;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
 import java.util.regex.Pattern;
 
 import static org.codehaus.groovy.ast.tools.GeneralUtils.*;
-import static org.grails.compiler.injection.GrailsASTUtils.applyDefaultMethodTarget;
-import static org.grails.compiler.injection.GrailsASTUtils.applyMethodTarget;
-import static org.grails.compiler.injection.GrailsASTUtils.buildGetMapExpression;
-import static org.grails.compiler.injection.GrailsASTUtils.buildGetPropertyExpression;
-import static org.grails.compiler.injection.GrailsASTUtils.buildSetPropertyExpression;
-import static org.grails.compiler.injection.GrailsASTUtils.hasAnnotation;
-import static org.grails.compiler.injection.GrailsASTUtils.hasParameters;
-import static org.grails.compiler.injection.GrailsASTUtils.isInheritedFromTrait;
-import static org.grails.compiler.injection.GrailsASTUtils.removeAnnotation;
+import static org.grails.compiler.injection.GrailsASTUtils.*;
 
 /**
  * Enhances controller classes by converting closures actions to method actions and binding
@@ -798,14 +786,12 @@ public class ControllerActionTransformer implements GrailsArtefactClassInjector,
             }
             
             if (argumentIsValidateable) {
-                final MethodCallExpression validateMethodCallExpression =
-                        new MethodCallExpression(new VariableExpression(paramName), "validate", EMPTY_TUPLE);
-                final MethodNode validateMethod =
-                        commandObjectNode.getMethod("validate", new Parameter[0]);
+                final MethodCallExpression validateMethodCallExpression = callX(localVarX(paramName, commandObjectNode), "validate");
+                final MethodNode validateMethod = commandObjectNode.getMethod("validate", new Parameter[0]);
                 if (validateMethod != null) {
                     validateMethodCallExpression.setMethodTarget(validateMethod);
                 }
-                final Statement ifCommandObjectIsNotNullThenValidate = new IfStatement(new BooleanExpression(new VariableExpression(paramName)), new ExpressionStatement(validateMethodCallExpression), new ExpressionStatement(new EmptyExpression()));
+                final Statement ifCommandObjectIsNotNullThenValidate = ifS(boolX(varX(paramName)), stmt(validateMethodCallExpression));
                 wrapper.addStatement(ifCommandObjectIsNotNullThenValidate);
             } else {
                 // try to dynamically invoke the .validate() method if it is available at runtime...

--- a/grails-test-suite-web/src/test/groovy/org/grails/web/commandobjects/CommandObjectsSpec.groovy
+++ b/grails-test-suite-web/src/test/groovy/org/grails/web/commandobjects/CommandObjectsSpec.groovy
@@ -403,7 +403,7 @@ class TestController {
 }
 
 
-class ParentCommand {
+class ParentCommand implements Validateable {
     int testId
 }
 

--- a/grails-test-suite-web/src/test/groovy/org/grails/web/commandobjects/CommandObjectsSpec.groovy
+++ b/grails-test-suite-web/src/test/groovy/org/grails/web/commandobjects/CommandObjectsSpec.groovy
@@ -5,7 +5,6 @@ import grails.testing.gorm.DataTest
 import grails.testing.web.controllers.ControllerUnitTest
 import grails.validation.Validateable
 import spock.lang.Issue
-import spock.lang.PendingFeature
 import spock.lang.Specification
 
 class CommandObjectsSpec extends Specification implements ControllerUnitTest<TestController>, DataTest {
@@ -323,7 +322,6 @@ class CommandObjectsSpec extends Specification implements ControllerUnitTest<Tes
         model.pId == 2
     }
 
-    @PendingFeature()
     @Issue('https://github.com/grails/grails-core/issues/13945')
     void "calling actions involving inherited command objects - parent command"() {
         given:

